### ISSUE TITLE
Low Success Rates

### ISSUE THREAD

**Issue by aozmekik** [kibok90/cvpr2018-hnd#2]
I have prepared all the taxonomy files and data. Although I checked all the files repeatedly, I couldn't got successful results in training. At least I need an idea of ​​where I might have made a mistake, because your last suggestions was very helpful.

**Success Graphs:**
![TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_test_acc](https://user-images.githubusercontent.com/32572999/62125810-db176600-b2d6-11e9-8595-b8c8e306e074.png)
![TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_val_acc](https://user-images.githubusercontent.com/32572999/62125811-db176600-b2d6-11e9-8856-5ef15686ab9d.png)

**Training Output:**
```
FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Namespace(batch_size=5, class_wise=True, cnn='resnet101', dataset='Minibar', ex_smooth=1.0, gpu=True, keep=False, known_guarantee=0.5, label_smooth=0.0, loo=1.0, lr=0.01, lr_decay=0.1, method='TD+LOO', no_last_relu=False, novel_score=False, num_epochs=50, num_layers=0, num_lr_decay=2, radius=-1, relabel=50, save_freq=100, save_results=False, softmax='l', td_name='TD_-1_1e+00_0e+00_1e-02_1e-02', test=True, test_data_norm=False, test_relu=True, wd=0.01, workers=0)
taxonomy for TD+LOO;    0.004 s
generate init weight: train/2048_18_td.pth
save init weight: train/2048_18_td.pth
generate init weight: train/18_19.pth
save init weight: train/18_19.pth
train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02
random order;    2.696 s
train data_loader;    2.606 s
val data_loader;    2.709 s
known data_loader;    2.768 s
novel data_loader;    2.775 s
   0/  50 e; start training;    2.775 s
   1/  50 e;   1/4044 i; lr: 1e-02; bl:     5.245; ml:     5.245;    4.140 s
   1/  50 e; 808/4044 i; lr: 1e-02; bl:     4.263; ml:     4.065;    7.029 s
   1/  50 e; 1616/4044 i; lr: 1e-02; bl:     3.997; ml:     3.935;    9.907 s
   1/  50 e; 2424/4044 i; lr: 1e-02; bl:     3.031; ml:     3.879;   12.803 s
   1/  50 e; 3232/4044 i; lr: 1e-02; bl:     4.339; ml:     3.843;   15.719 s
   1/  50 e; 4040/4044 i; lr: 1e-02; bl:     4.115; ml:     3.825;   18.574 s
   1/  50 e; 4044/4044 i; lr: 1e-02; bl:     3.144; ml:     3.825;   18.589 s
   1/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_1.pth;   18.591 s
   2/  50 e;   1/4044 i; lr: 1e-02; bl:     3.849; ml:     3.849;   18.594 s
   2/  50 e; 808/4044 i; lr: 1e-02; bl:     3.427; ml:     3.720;   21.437 s
   2/  50 e; 1616/4044 i; lr: 1e-02; bl:     4.653; ml:     3.739;   24.299 s
   2/  50 e; 2424/4044 i; lr: 1e-02; bl:     2.939; ml:     3.747;   27.182 s
   2/  50 e; 3232/4044 i; lr: 1e-02; bl:     3.546; ml:     3.742;   30.071 s
   2/  50 e; 4040/4044 i; lr: 1e-02; bl:     4.228; ml:     3.740;   32.960 s
   2/  50 e; 4044/4044 i; lr: 1e-02; bl:     3.898; ml:     3.740;   32.975 s
   2/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_2.pth;   32.977 s
   2/  50 e; remove train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_1.pth;   32.977 s
   3/  50 e;   1/4044 i; lr: 1e-02; bl:     3.704; ml:     3.704;   32.980 s
   3/  50 e; 808/4044 i; lr: 1e-02; bl:     2.874; ml:     3.730;   35.858 s
   3/  50 e; 1616/4044 i; lr: 1e-02; bl:     4.374; ml:     3.744;   38.738 s
   3/  50 e; 2424/4044 i; lr: 1e-02; bl:     3.790; ml:     3.731;   41.610 s
   3/  50 e; 3232/4044 i; lr: 1e-02; bl:     3.541; ml:     3.739;   44.487 s
   3/  50 e; 4040/4044 i; lr: 1e-02; bl:     3.019; ml:     3.738;   47.489 s
   3/  50 e; 4044/4044 i; lr: 1e-02; bl:     3.735; ml:     3.738;   47.505 s
   3/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_3.pth;   47.506 s
   3/  50 e; remove train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_2.pth;   47.506 s
   4/  50 e;   1/4044 i; lr: 1e-03; bl:     4.119; ml:     4.119;   47.510 s
   4/  50 e; 808/4044 i; lr: 1e-03; bl:     3.110; ml:     3.517;   50.400 s
   4/  50 e; 1616/4044 i; lr: 1e-03; bl:     3.362; ml:     3.525;   53.306 s
   4/  50 e; 2424/4044 i; lr: 1e-03; bl:     3.830; ml:     3.530;   56.231 s
   4/  50 e; 3232/4044 i; lr: 1e-03; bl:     3.668; ml:     3.530;   59.115 s
   4/  50 e; 4040/4044 i; lr: 1e-03; bl:     3.083; ml:     3.531;   62.009 s
   4/  50 e; 4044/4044 i; lr: 1e-03; bl:     3.516; ml:     3.531;   62.022 s
   4/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_4.pth;   62.024 s
   4/  50 e; remove train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_3.pth;   62.024 s
   5/  50 e;   1/4044 i; lr: 1e-03; bl:     3.505; ml:     3.505;   62.028 s
   5/  50 e; 808/4044 i; lr: 1e-03; bl:     3.277; ml:     3.513;   64.883 s
   5/  50 e; 1616/4044 i; lr: 1e-03; bl:     3.397; ml:     3.535;   67.748 s
   5/  50 e; 2424/4044 i; lr: 1e-03; bl:     3.403; ml:     3.539;   70.642 s
   5/  50 e; 3232/4044 i; lr: 1e-03; bl:     3.238; ml:     3.537;   73.523 s
   5/  50 e; 4040/4044 i; lr: 1e-03; bl:     3.684; ml:     3.534;   76.438 s
   5/  50 e; 4044/4044 i; lr: 1e-03; bl:     3.605; ml:     3.534;   76.453 s
   5/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_5.pth;   76.454 s
   5/  50 e; remove train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_4.pth;   76.454 s
   6/  50 e;   1/4044 i; lr: 1e-04; bl:     3.225; ml:     3.225;   76.458 s
   6/  50 e; 808/4044 i; lr: 1e-04; bl:     4.176; ml:     3.522;   79.524 s
   6/  50 e; 1616/4044 i; lr: 1e-04; bl:     3.747; ml:     3.510;   82.489 s
   6/  50 e; 2424/4044 i; lr: 1e-04; bl:     3.805; ml:     3.509;   85.509 s
   6/  50 e; 3232/4044 i; lr: 1e-04; bl:     2.742; ml:     3.514;   88.789 s
   6/  50 e; 4040/4044 i; lr: 1e-04; bl:     3.738; ml:     3.514;   91.996 s
   6/  50 e; 4044/4044 i; lr: 1e-04; bl:     3.472; ml:     3.514;   92.011 s
   6/  50 e; save train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_6.pth;   92.012 s
   6/  50 e; remove train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02_model_5.pth;   92.012 s
   7/  50 e; training done;   92.012 s
train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02
val;  100.790 s
bias: -16.0000; acc known: 31.71; novel:  0.00; auc  : 14.82; 
train/Minibar/resnet101/TD+LOO/TD_-1_1e+00_0e+00_1e-02_1e-02/TD+LOO_1e+00_0e+00_cw_relu_l_1e-02_1e-02
test;  101.473 s
bias: -16.0000; acc known: 31.71; novel:  0.00; auc  :  0.06; 

```
**Custom Files and Data:** 
[Github Link](https://github.com/drh0use1/minibar-taxonomy)

 Thanks.

**Comment by kibok90** [comment]
Hi Ahmed,

Sorry for confusion, it seems I didn't put a detail about TD+LOO in this github. I think I need to update this github when I have some time.
The current version of code does not train TD+LOO end-to-end, but train LOO on top of the output of a pretrained TD model. In your log, the code couldn't load any pretrained TD model, so I think the result you see in the log is the case when the TD model has randomly initialized and fixed parameters.

So, my suggestion is to remove the trained TD+LOO model (otherwise my code loads any existing model), train a TD model first, and then train a TD+LOO model.
Then you will see a different log and hopefully it works well.

By the way, I recommend to try Relabel or LOO method first. If the performance is not good in these simpler methods, then the semantic information in your taxonomy is either 1) not well-aligned with images or 2) just too small. Since you have a small number of child nodes (18) and classes (19), I guess your taxonomy is somewhat simple/shallow. In this case, based on my observation on CIFAR-100 (see Appendix A.3 in my paper), the performance gain from TD features (TD+LOO vs. LOO) would be marginal or even slightly worse.

Also, you are using a new dataset, so you may need to find better hyperparameters. If you have different hyperparameters for TD, then you should set an appropriate `-tdname` when training TD+LOO.

**Comment by aozmekik** [comment]
Thank you very much for your quick response. 

**The test results are as follows:**
![photo5915612612203229286](https://user-images.githubusercontent.com/32572999/62296619-261caf00-b478-11e9-8864-4eb516da7036.jpg)
![photo5915612612203229287](https://user-images.githubusercontent.com/32572999/62296620-26b54580-b478-11e9-9e23-d36c26010286.jpg)





Is there anything you'd like to recommend to make these results better?

**Comment by kibok90** [comment]
If I were you, I would first analyze why the novel class accuracy is bounded by a number around 0.5. Your taxonomy is small, so its visualization might be helpful.

**Comment by maggieHao** [comment]
Hiya, I used the CUB dataset and try to reproduce your results on a different version of pytorch. The results of methods TD, TD+LOO, and LOO are the same as yours. However, the result of Relabel is quite low: val "bias: 11.6333; acc known: 50.00; novel:  1.87; auc  :  1.83; " and test "bias:  9.5543; acc known: 50.00; novel:  5.16; auc  :  3.56;"

So I was wondering what's wrong with it?

**Comment by kibok90** [comment]
Hi,

This repository is not guaranteed work well with a different version of libraries.
If only the performance Relabel is low, you may retry it after removing the model files and caches. Relabel generates caches in `{dataset_path}/relabels_{:d}.h5`.
`utils.relabel()` would be the only function uniquely used for Relabel (and generating some caches), so you may also want to check this:
https://github.com/kibok90/cvpr2018-hnd/blob/e5fd01043146bf158283e096f54aef1448a1f3fb/utils.py#L112-L162
